src/frontend/src/optimizer/plan_node/stream_table_scan.rs

```diff
@@ -282,16 +302,19 @@ impl StreamTableScan {
             })
             .collect_vec();
 
-        // TODO: snapshot read of upstream mview
-        let batch_plan_node = BatchPlanNode {
-            table_desc: Some(self.core.table_desc.to_protobuf()),
-            column_ids: upstream_column_ids.clone(),
+        // This refers to the output indices of the originating stream.
+        let output_indices_origin = self.core.output_and_pk_column_indices();
+        let upstream_table_catalog = self
+            .get_upstream_state_table()
+            .clone()
+            .into_replicated(output_indices_origin);
+
+        let arrangement_table = if self.stream_scan_type == StreamScanType::ArrangementBackfill {
+            Some(upstream_table_catalog.to_internal_table_prost())
+        } else {
+            None
         };
```

Comment:
According to these lines, `output_indices_origin` is used by `ArrangementBackfill` only. What do you mean it is also used by the no shuffle backfill?

Reply:
I mean storage table has a similar concept as well:
```
    /// Create a  [`StorageTableInner`] given a complete set of `columns` and a partial
    /// set of `column_ids`. The output will only contains columns with the given ids in the same
    /// order.
    #[allow(clippy::too_many_arguments)]
    pub fn new_partial(
        store: S,
        table_id: TableId,
        table_columns: Vec<ColumnDesc>,
        column_ids: Vec<ColumnId>,
        order_types: Vec<OrderType>,
        pk_indices: Vec<usize>,
        distribution: Distribution,
        table_options: TableOption,
        value_indices: Vec<usize>,
        read_prefix_len_hint: usize,
        versioned: bool,
    ) -> Self {
```

Where it uses `column_ids` to only read a specific set of columns.

The usage is similar to output_indices.

Reply:
Which is what `no_shuffle_backfill` uses

Reply:
My point is that `TableCatalog` is a struct used more than just for arrangement backfill. Adding an additional field just for arrangement backfill seems overkill. As you mentioned, for non-shuffle backfill `StreamScanNode` has a separate field `upstream_column_ids` to represent the ouput_indices, can we do that for the arrangement backfill in the same way? That would make `StreamScanNode` meta data more consistent between arrangement and no shuffle backfill.

Reply:
> My point is that `TableCatalog` is a struct used more than just for arrangement backfill. Adding an additional field just for arrangement backfill seems overkill. As you mentioned, for non-shuffle backfill `StreamScanNode` has a separate field `upstream_column_ids` to represent the ouput_indices, can we do that for the arrangement backfill in the same way? That would make `StreamScanNode` meta data more consistent between arrangement and no shuffle backfill. 

Alright, let me try to unify them.

Reply:
Including output_indices in the table catalog made it a lot more complicated than it had to be. 

I have now fixed it. Thanks for the great suggestion.